src/Common/src/TypeSystem/Interop/IL/Marshaller.cs

```diff
@@ -69,36 +69,165 @@ abstract class Marshaller
         public MarshallerKind MarshallerKind;
         public MarshallerType MarshallerType;
         public MarshallerKind ElementMarshallerKind;
+        public TypeDesc ManagedType;
         public TypeDesc ManagedParameterType;
         protected Marshaller[] Marshallers;
-        private TypeDesc _nativeParameterType;
+        private TypeDesc _nativeType;
+        private TypeDesc _nativeParamType;
 
+        /// <summary>
+        /// Native Type of the value being marshalled
+        /// For by-ref scenarios (ref T), Native Type is T
+        /// </summary>
+        public TypeDesc NativeType
+        {
+            get
+            {
+                if (_nativeType == null)
+                {
+                    _nativeType = GetNativeTypeFromMarshallerKind(
+                        ManagedType,
+                        MarshallerKind, ElementMarshallerKind,
+                        PInvokeParameterMetadata.MarshalAsDescriptor);
+                    Debug.Assert(_nativeType != null);
+                }
+
+                return _nativeType;
+            }
+        }
+
+        /// <summary>
+        /// NativeType appears in function parameters
+        /// For by-ref scenarios (ref T), NativeParameterType is T*
+        /// </summary>
         public TypeDesc NativeParameterType
         {
             get
             {
-                if (_nativeParameterType == null)
+                if (_nativeParamType == null)
                 {
-                    _nativeParameterType = GetNativeTypeFromMarshallerKind(ManagedParameterType,
-                                                        MarshallerKind, ElementMarshallerKind,
-                                                    PInvokeParameterMetadata.MarshalAsDescriptor);
-                    Debug.Assert(_nativeParameterType != null);
+                    TypeDesc nativeParamType = NativeType;
+                    if (IsNativeByRef)
+                        nativeParamType = nativeParamType.MakePointerType();
+                    _nativeParamType = nativeParamType;
                 }
-                return _nativeParameterType;
+
+                return _nativeParamType;
             }
         }
 
         public bool In;
         public bool Out;
         public bool Return;
         public bool Optional;
-        public bool IsByRef;
+        public bool IsManagedByRef;                     // Whether managed argument is passed by ref
+        public bool IsNativeByRef;                      // Whether native argument is passed by byref
+                                                        // There are special cases (such as LpStruct, and class) that 
+                                                        // isNativeByRef != IsManagedByRef
         public MarshalDirection MarshalDirection;
         protected PInvokeILCodeStreams _ilCodeStreams;
-        protected ILLocalVariable _vManaged;
-        protected ILLocalVariable _vNative;
+        protected Home _vManaged;
```

Comment:
Nit: Can you please rename this to something like _managedHome and also _vNative to _nativeHome.  "_v" prefix I suppose was for ILLocal**V**ariables?